staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -210,16 +210,82 @@ func (c *controller) processLoop() {
 //     it will get an object of type DeletedFinalStateUnknown. This can
 //     happen if the watch is closed and misses the delete event and we don't
 //     notice the deletion until the subsequent re-list.
+//
+// Implementers MUST ALSO implement OnAdd, from either
+// ResourceEventHandlerInspecific or ResourceEventHandlerDetailed (preferred).
+//
+// REQUIRED ACTION NOTICE: `OnAdd(obj interface{}, isInInitialList bool)` will
+// be added to this interface in v1.28.
```

Comment:
If we're going to change the interface, I don't see why we would wait a release and make this interface less usable in the meantime. If changing this interface now has a lot of ripples and is painful to do in this PR, that might be a sign that we should ease that transition, rather than simply delay it.

If we're going to change the signature of OnAdd, I think it would be better to:
1. change the interface now unconditionally
2. leave the member fields in ResourceEventHandlerFuncs as-is (so ResourceEventHandlerFuncs#OnAdd would just drop the isInInitialList param)
3. define ResourceEventHandlerDetailedFuncs type as you did in this PR, so people who want the detailed information can get it

Trying out that approach only required changing one additional OnAdd implementation (in test/integration/client/exec_test.go), because almost all callers use ResourceEventHandlerFuncs

It also dropped several runtime error paths where something that didn't implement an acceptable OnAdd function could get passed around.

Reply:
OK I pushed this as a new commit. I think it risks more clients having to do something, but it is shorter

> It also dropped several runtime error paths

I think they are replaced by much subtler bugs of failing to propagate isInInitialList, if any middleware exists out there. Bugs would have been caught at initialization time.